nemo_aligner/algorithms/reinforce.py

```diff
@@ -0,0 +1,612 @@
+            else:
+                pad_value = eos_id if k == "response_tokens" else 0
+
+                list_of_tensors = [row.flatten() for tensor in list_of_tensors for row in tensor]
+                # TODO: can we avoid padding locally then padding globally?
+                tensor = torch.nn.utils.rnn.pad_sequence(list_of_tensors, batch_first=True, padding_value=pad_value)
+
+                # find the max sequence length globally
+                max_seqlen = torch.tensor([tensor.size(-1)], dtype=torch.long, device=torch.cuda.current_device())
+                torch.distributed.all_reduce(max_seqlen, op=torch.distributed.ReduceOp.MAX)
+
+                if rollout_batch_seq_length is None or max_seqlen >= rollout_batch_seq_length:
+                    pad_seq_len = max_seqlen.item()
+                else:
+                    # response tokens must be B x S because computing log probs requires us to offset by 1
+                    pad_seq_len = rollout_batch_seq_length if k == "response_tokens" else rollout_batch_seq_length - 1
+
+                tensor = torch.nn.functional.pad(tensor, (0, pad_seq_len - tensor.size(-1)), value=pad_value)
+
+            stacked_dict[k] = tensor
+
+        return stacked_dict
+
+    def gather_and_balance_globally(self):
+        global_rollout_batch = type(self)()
+
+        for k, tensor in self.data.items():
+            # with reshard enabled, PP groups turn into DP groups. So need to balance them first and then
+            # balance by all the original DP groups
+            # NOTE: this logic needs to use the pure parallel state, that is one without sharding but needs
+            # to ping the is_trt_llm_reshard variable
+            if is_trt_llm_reshard():
+                tensor = rebalance_nd_tensor(tensor, group=mcore_parallel_state.get_pipeline_model_parallel_group())
+
+            tensor = rebalance_nd_tensor(tensor, group=mcore_parallel_state.get_data_parallel_group())
+            global_rollout_batch[k] = tensor
+
+        return global_rollout_batch
+
+    def chunk(self, rank, split_size, seed):
+        chunked_rollout_batch = type(self)()
+
+        batch_set = set(tensor.size(0) for tensor in self.data.values())
+        assert len(batch_set) == 1, "batch sizes are not the same across the rollout batch"
+        B = batch_set.pop()
+
+        g_cpu = torch.Generator()
+        g_cpu.manual_seed(seed)
+        indices = torch.arange(B)
+
+        for k in self.data:
+            chunked_rollout_batch[k] = self.data[k][indices].clone()
+
+        return chunked_rollout_batch
+
+
+def compute_num_rollout_microbatches(dataloader):
+    return divide(
+        divide(dataloader.batch_sampler.global_batch_size, dataloader.batch_sampler.micro_batch_size),
+        parallel_state.get_data_parallel_world_size(),
+    )
+
+
+class ReinforceTrainer:
+    """Trainer to coordinate REINFORCE training
+    """
+
+    def __init__(
+        self,
+        cfg: DictConfig,
+        model,
```

Comment:
I think you should statically type at least the model. It helps folks who are using editors to jump to definitions

Reply:
Fixed this.